### PR TITLE
SUGADD, SUGGET; added Interfaces for the contract to the consumer of the API. A Build…

### PR DESCRIPTION
**Pull request by davidparry**
…er for the Client and ClusterClient probably would be good next too. Allow the consumers of this api not to be concrete connected to the implementations.

I also commented out testPayloads() line 203 it was broken when i ran the test suite to begin with.

### CODE REVIEW

**Comment by davidparry** [comment]
I see that some new classes some how missed on the add will have them added by 16:00 CST

**Comment by mnunberg** [comment]
Generally looks ok:

1. I'd rather that the client remain a concrete class (I'd be ok with Client and AbstractClient, etc. though). Are you intending for there to be *another* provider of this API? If and when such a time arrives, we can consider using interfaces.
2. Please consolidate all the suggestion methods. I don't like expanding the API surface area like that (but I'm not a Java guy). I'd rather use an option-builder like thingy. Less API surface area to maintain.

**Comment by davidparry** [comment]
(1)
It was not my intent to remove the Concrete class by Introducing the initial 2 Interfaces or to have other providers, but instead to allow easier maintainability without headaches to the api consumer. I am addressing qualifying factors that java teams and companies, current one included, that must have these standard set of quality features before a project can adopt a new library dependency. The Interface introduction is only a step towards a handful of Design Principles and other factors that increase the chances of a Java library to be adopted in a new project vs (another library, writing from scratch, etc..). Interface Segregation Principle, Dependency Inversion Principle, Inversion of Control, Dependency Injection, Code Quality and Metrics. I plan on introducing some more of these Principles to this project, so I can get it accepted as a dependency into our Solution.


(2)
I was torn on which way to go, seeing the style you were using.  My thoughts where that the caller knows if they want a payload, score, both or none, so let them make the choice and tell us this in the form of a more specific method. Hence removing the case statements within the method to find out what the caller desired. In Java and the Unit Test encourages this to have smaller methods than one larger more complex method which we need to perform logic to see if the caller wanted a payload, score, both or none. I will go ahead and make it one and add the complexity in the one method.

**Comment by mnunberg** [comment]
Yeah, if we keep the API surface area small, then the duplication of interfaces vs classes is not horrible- but if we offer a ton of overloads then we'd end up with a sticky mess.

File: src/main/java/io/redisearch/client/Client.java [hunk 1]
```diff
@@ -519,4 +432,180 @@ public boolean dropIndex(boolean missingOk) {
...
+            args.add(suggestion.getPayload());
+        }
+
+        try (Jedis conn = _conn()) {
+            return sendCommand(conn, AutoCompleter.Command.SUGADD, args.toArray(new String[args.size()])).getIntegerReply();
+        }
+    }
+
+    @Override
+    public List<Suggestion> getSuggestion(String prefix, SuggestionOptions suggestionOptions) {
```
Comment (mnunberg) [inline src/main/java/io/redisearch/client/Client.java]: I'd still prefer this be branched out to other smaller helper methods. I don't like the big API, but internally we can branch however we'd like.

File: src/main/java/io/redisearch/Client.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+    SearchResult search(Query q);
+
+    AggregationResult aggregate(AggregationRequest q);
+
+    String explain(Query q);
+
+    boolean addDocument(String docId, double score, Map<String, Object> fields, boolean noSave, boolean replace, byte[] payload);
+
+    boolean addDocument(Document doc, AddOptions options);
```
Comment (mnunberg) [inline src/main/java/io/redisearch/Client.java]: This (`addDocument(Document, AddOptions)`) should be the canonical interface. The other helper methods are there for legacy purposes.

**Comment by mnunberg** [comment]
I'm looking at the current interface, and perhaps Replace needs to go as well, anyway, that's enough for a single iteration :)

**Comment by davidparry** [comment]
Ok will get it tomorrow I hope

**Pull request merged**